Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,556 @@
+
+```swift
+/// The base LambdaHandler protocol
+public protocol StreamingLambdaHandler: ~Copyable {
+    /// The business logic of the Lambda function
+    /// - Parameters:
+    ///   - event: The invocation's input data
+    ///   - responseWriter: A ``LambdaResponseWriter`` to write the invocations response to.
+    ///                     If no response or error is written to the `responseWriter` it will
+    ///                     report an error to the invoker.
+    ///   - context: The LambdaContext containing the invocation's metadata
+    mutating func handle(_ event: ByteBuffer, responseWriter: consuming some LambdaResponseWriter, context: LambdaContext) async
+}
+```
+
+The protocol is defined in a way that supports a broad range of use-cases. The handle method is marked as `mutating` to allow handlers to be implemented with a `struct`. Users can opt to even mark the handler as non-copyable.
+
+An implementation that sends the number 1 to 10 every 500ms could look like this:
+
+```swift
+struct SendNumbersWithPause: StreamingLambdaHandler {
+    func handle(
+        _ event: ByteBuffer,
+        responseWriter: consuming some LambdaResponseWriter,
+        context: LambdaContext
+    ) async {
+        for i in 1...10 {
+            responseWriter.write(ByteBuffer(string: #"\#(i)\n\r"#))
+            try? await Task.sleep(for: .milliseconds(500))
+        }
+        responseWriter.finish()
+    }
+}
+```
+
+The method is not marked as `throws` as we want the users of this API to handle errors themselves and call `LambdaResponseWriter.reportError(_:)` if needed.
+
+### LambdaRuntime
+
+`LambdaRuntime` is the class that communicates with the Lambda control plane as defined in [Building a custom runtime for AWS Lambda](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-custom.html) and forward the invocations to the provided `StreamingLambdaHandler`. It will conform to `ServiceLifecycle.Service` to provide support for `swift-service-lifecycle`.
+
+```swift
+/// The LambdaRuntime object. This object communicates with the Lambda control plane
+/// to fetch work and report errors.
+public final class LambdaRuntime<Handler>: ServiceLifecycle.Service, Sendable
+    where Handler: StreamingLambdaHandler, Handler: ~Copyable
+{
+
+    /// Create a LambdaRuntime by passing a handler, an eventLoop and a logger.
+    /// - Parameter handler: A ``StreamingLambdaHandler`` that will be invoked
+    /// - Parameter eventLoop: An ``EventLoop`` on which the LambdaRuntime will be
+    ///                        executed. Defaults to an EventLoop from
+    ///                        ``NIOSingletons.posixEventLoopGroup``.
+    /// - Parameter logger: A logger
+    public init(
+        handler: consuming sending Handler,
+        eventLoop: EventLoop = Lambda.defaultEventLoop,
+        logger: Logger = Logger(label: "Lambda")
```

Comment:
IMO we shouldn't default the logger. Let users pass this in and if they really want to create a fresh logger they can decide to do that

Reply:
I think that makes sense. Would you default the Logger to nil then? The idea of using a default logger here is that new adopters don't need to learn about Logging first.

Reply:
Yeah I would default to `nil` and basically have no logging if nothing is passed.

Reply:
I'd disagree with this. Most people (I would assume?) will be using the lambda handler directly rather than via another framework like Hummingbird or Vapor. It would be weird to deploy the simplest thing and get _no_ logging by default as that's the first place people check to see if things are working or not.

Reply:
+1 to @0xTim's point also

Reply:
@FranzBusch Can you get behind the arguments and rest your case?

Reply:
The fact that I propose to default the proposed task local logger to something is making it clear to me that this should also default to a standard logger. In the future, if we do the task local we can just default this logger to the task local value.

Reply:
+1 on Tim's comment